src/subscription_manager/cli_command/cli.py

```diff
@@ -322,17 +322,15 @@ def main(self, args: Optional[List[str]] = None) -> Optional[int]:
             )
             config_changed = True
 
-        # support foo.example.com:3128 format
         if hasattr(self.options, "proxy_url") and self.options.proxy_url:
-            parts = remove_scheme(self.options.proxy_url).split(":")
-            self.proxy_hostname = parts[0]
-            # no ':'
-            if len(parts) > 1:
-                self.proxy_port = int(parts[1])
-            else:
-                # if no port specified, use the one from the config, or fallback to the default
-                self.proxy_port = conf["server"].get_int("proxy_port") or rhsm.config.DEFAULT_PROXY_PORT
-            config_changed = True
+            default_proxy_port = conf["server"].get_int("proxy_port") or rhsm.config.DEFAULT_PROXY_PORT
+            proxy_user, proxy_pass, proxy_hostname, proxy_port, proxy_prefix = rhsm.utils.parse_url(
+                self.options.proxy_url, default_port=default_proxy_port
+            )
+            self.proxy_user = proxy_user
+            self.proxy_password = proxy_pass
```

Comment:
this enables username & password for a proxy server to be specified via `--proxy`, whereas they were ignored before this change; considering there are the `--proxyuser` & `--proxypassword` option, and those will still override those attributes; for example:
```
subscription-manager --proxy proxy-user:proxy-password@proxy-host:port --proxyuser new-user
```
this will actually try to use the `proxy-host` with username `new-user` and password `proxy-password`, and `proxy-user` is silently discarded, which I do not think it is good for the user (they specify an option, or part of it, which gets overridden by another option)

I'd remove this, and keep `--proxy` to a simple hostname+port definition.

Reply:
I think that current behavior is correct for two reasons.

1. It does not break old behavior, when user and password were ignored in URL.
2. When you use `--proxyuser` then you explicitly say that you want to use specified user. This desire is IMHO more explicit specifying it using `--proxy URL`.

If your proposal was implemented, then somebody could argue that `--proxyuser` is overridden by `--proxy` and it is breaking old behavior.

Reply:
> If your proposal was implemented

I guess I explained myself badly: my proposal here is to **not** pick username & password from `--proxy`, retaining the current behaviour.

Reply:
Why not to pick username & password from `--proxy`, when this CLI option is used and it contains username & password? I think that we should parse the URL properly according specification.

Reply:
I understand the reasons about proper parsing, and normally I'd agree with that. However, the challenge here is that, even if small, this is a behaviour change: suddenly parts of `--proxy` are used whereas they were not previously.

Since the rest of the changes are good, what do you think about keep the existing behaviour of the `--proxy` parsing in this PR, and discuss the usage of credentials from `--proxy` as a separate change? This way we can discuss that with QE to check for the impact.

Reply:
I'm asking to split the behaviour change for `--proxy` in its separate PR because, as we both clearly see, we do not agree on having this change in, while the actual fix (parsing IPv6 addresses correctly) is good and can be merge (and it would had been merged already long ago, if not for the behaviour change).

So I'll reiterate my request: can you please move the `--proxy` behaviour change away from this PR to a new one, so we can merge this? Thanks in advance!

Reply:
The change for `--proxy` just cannot be split away. If it was split away, then it would not fix the bug mentioned in the first comment. I believe that we want to fix bugs.

I will retrace what is content of this PR. There were two ways how to parse URLs. Both were wrong. First commit fixes parsing of URLs from `rhsm.conf`. Second commit just uses fixed `parse_url()` (fixed in first commit) for `--proxy` CLI option. Last commit fixes only minor things related to debug print of traffic.

I would like to clarify my previous comments. I want to have this change in. I do not see any reason, why this change should not be in. I still believe that we want to fix bugs.

Reply:
> The change for `--proxy` just cannot be split away.

Sure it can: it's about removing these two added lines:
```python
            self.proxy_user = proxy_user
            self.proxy_password = proxy_pass
```
This is the behaviour change I'm pointing out, and I'd like to see discussed and handled **separately**.

> Second commit just uses fixed `parse_url()` (fixed in first commit) for `--proxy` CLI option.

Yes, exactly, and I don't see where it says "change the behaviour of `--proxy`" as part of that, since the parsing issue can be fixed without the behaviour change.

The bug fix was "parsing IPv6 addresses properly", and this is a behaviour change that was never discussed earlier and it is not even required to fix the parsing of IPv6 addresses.

I want the change of this bug fix in already as well. Without this behaviour change, I would have approved and merged this few weeks ago already.

Reply:
Let me clarify one thing. Is your concern following? Current version of `subscription-manager` ignores `proxy_username` and `proxy_password` in URL of following command line option `--proxy="https://proxy_username:proxy_password@proxy.company.com:3128"` and there is almost zero probability that somebody uses such URL, because username and password are ignored. If somebody needs to use proxy_username and proxy_password, then they use `--proxyuser` and `--proxypassword` command line options, right?

If this PR was merged, then `proxy_username` and `proxy_password` of `--proxy="https://proxy_username:proxy_password@proxy.company.com:3128"` would not be ignored, but the username & password would not be ignored only in the case, when `--proxyuser` and `--proxypassword` were not provided, because `--proxyuser` and `--proxypassword` still has higher priority over `proxy_username` and `proxy_password` of `--proxy="https://proxy_username:proxy_password@proxy.company.com:3128"`.